src/main/java/ch/njol/skript/patterns/Keyword.java

```diff
@@ -42,57 +49,49 @@ abstract class Keyword {
 	 * @param first The pattern to build keywords from.
 	 * @return A list of all keywords within <b>first</b>.
 	 */
+	@Contract("_ -> new")
 	public static Keyword[] buildKeywords(PatternElement first) {
+		return buildKeywords(first, true, 0);
+	}
+
+	/**
+	 * Builds a list of keywords starting from the provided pattern element.
+	 * @param first The pattern to build keywords from.
+	 * @param starting Whether this is the start of a pattern.
+	 * @return A list of all keywords within <b>first</b>.
+	 */
+	@Contract("_, _, _ -> new")
+	private static Keyword[] buildKeywords(PatternElement first, boolean starting, int depth) {
 		List<Keyword> keywords = new ArrayList<>();
 		PatternElement next = first;
-		boolean starting = true; // whether it is the start of the pattern
-		boolean ending = next.next == null; // whether it is the end of the pattern
 		while (next != null) {
 			if (next instanceof LiteralPatternElement) { // simple literal strings are keywords
 				String literal = next.toString().trim();
 				while (literal.contains("  "))
 					literal = literal.replace("  ", " ");
-				keywords.add(new SimpleKeyword(literal, starting, ending));
-			} else if (next instanceof ChoicePatternElement) { // this element might contain some keywords
-				List<PatternElement> choiceElements = flatten(next);
-				if (choiceElements.stream().allMatch(e -> e instanceof LiteralPatternElement)) {
-					// all elements are literals, and this is a choice, meaning one of them must be required
-					// thus, we build a keyword that requires one of them to be present.
-					List<String> groupKeywords = choiceElements.stream()
-						.map(e -> {
-							String literal = e.toString().trim();
-							while (literal.contains("  "))
-								literal = literal.replace("  ", " ");
-							return literal;
-						})
-						.collect(Collectors.toList());
-					keywords.add(new GroupKeyword(groupKeywords, starting, ending));
-				}
-			} else if (next instanceof GroupPatternElement) { // groups need to be unwrapped (they might contain choices)
-				next = ((GroupPatternElement) next).getPatternElement();
-				continue;
+				if (!literal.isEmpty()) // empty string is not useful
+					keywords.add(new SimpleKeyword(literal, starting, next.next == null));
+			} else if (depth <= 1 && next instanceof ChoicePatternElement) { // attempt to build keywords from choices
+				final boolean finalStarting = starting;
+				final int finalDepth = depth;
+				// build the keywords for each choice
+				Set<Set<Keyword>> choices = ((ChoicePatternElement) next).getPatternElements().stream()
+					.map(element -> buildKeywords(element, finalStarting, finalDepth))
+					.map(ImmutableSet::copyOf)
+					.collect(Collectors.toSet());
+				if (choices.stream().noneMatch(Collection::isEmpty)) // each choice must have a keyword for this to work
```

Comment:
This looks kind of messy and inefficient, especially all the repeated collection copying. Maybe it would be faster to do it properly?

Reply:
this is unlikely to be an issue as this is a one-time execution when the pattern is compiled.